### PR TITLE
[krel] Initial commit krel ci-build command

### PR DESCRIPTION
**Pull request by justaugustus**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Initial commit of krel ci-build

`krel ci-build` is a replacement for the 'kubernetes_build' bootstrap
scenario.

The command is decomposed from https://github.com/kubernetes/test-infra/blob/7473f331fb69b1f88a2b846e5b8160c09ff07943/scenarios/kubernetes_build.py.

Signed-off-by: Stephen Augustus <saugustus@vmware.com>

ref: https://github.com/kubernetes/release/issues/1693

#### Does this PR introduce a user-facing change?

```release-note
[krel] Initial commit krel ci-build command
```

### CODE REVIEW

**Comment by justaugustus** [comment]
Still working on this but here are some initial logs from https://github.com/kubernetes/release/pull/1698/commits/61e3b618c826786e01aa01125923e1d7ccfd2ec6:

```console
$ time krel ci-build --fast --bucket k8s-release-kubernetes --docker-registry k8s-staging-kubernetes
```

```console
INFO Current working directory: /home/augustus/go/src/k8s.io/kubernetes 
INFO Getting workspace status                     
STABLE_BUILD_GIT_COMMIT 2570a9d94ced1954edfe4cbf3b84200f126ab717
STABLE_BUILD_SCM_STATUS clean
STABLE_BUILD_SCM_REVISION v1.20.0-beta.1.257+2570a9d94ced19
STABLE_BUILD_MAJOR_VERSION 1
STABLE_BUILD_MINOR_VERSION 20+
STABLE_DOCKER_TAG v1.20.0-beta.1.257_2570a9d94ced19
STABLE_DOCKER_REGISTRY k8s.gcr.io
STABLE_DOCKER_PUSH_REGISTRY staging-k8s.gcr.io
gitCommit 2570a9d94ced1954edfe4cbf3b84200f126ab717
gitTreeState clean
gitVersion v1.20.0-beta.1.257+2570a9d94ced19
gitMajor 1
gitMinor 20+
buildDate 2020-11-08T09:20:24Z
INFO Found workspace version: v1.20.0-beta.1.257+2570a9d94ced19 
INFO Checking if GCS build path (gs://k8s-release-kubernetes/ci/fast/v1.20.0-beta.1.257+2570a9d94ced19) exists 
INFO Checking if GCS build path (gs://k8s-release-kubernetes/ci/fast/v1.20.0-beta.1.257+2570a9d94ced19/kubernetes.tar.gz) exists 
INFO Checking if GCS build path (gs://k8s-release-kubernetes/ci/fast/v1.20.0-beta.1.257+2570a9d94ced19/bin) exists 
INFO The following error(s) occurred while looking for a build: [command /usr/bin/gsutil ls gs://k8s-release-kubernetes/ci/fast/v1.20.0-beta.1.257+2570a9d94ced19 did not succeed: BucketNotFoundException: 404 gs://k8s-release-kubernetes bucket does not exist.
 command /usr/bin/gsutil ls gs://k8s-release-kubernetes/ci/fast/v1.20.0-beta.1.257+2570a9d94ced19/kubernetes.tar.gz did not succeed: BucketNotFoundException: 404 gs://k8s-release-kubernetes bucket does not exist.
 command /usr/bin/gsutil ls gs://k8s-release-kubernetes/ci/fast/v1.20.0-beta.1.257+2570a9d94ced19/bin did not succeed: BucketNotFoundException: 404 gs://k8s-release-kubernetes bucket does not exist.
] 
INFO Proceeding with a new build...               
+++ [1108 04:20:32] Verifying Prerequisites....
+++ [1108 04:20:46] Removing _output directory
+++ [1108 04:20:47] Verifying Prerequisites....
+++ [1108 04:20:47] Building Docker image kube-build:build-d82df4bf66-5-v1.15.2-1
+++ [1108 04:20:51] Creating data container kube-build-data-d82df4bf66-5-v1.15.2-1
+++ [1108 04:21:10] Syncing sources to container
+++ [1108 04:21:15] Running build command...
stat /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/bazel-bin: no such file or directory
+++ [1108 04:21:17] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
+++ [1108 04:21:20] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
+++ [1108 04:21:24] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
+++ [1108 04:21:31] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
+++ [1108 04:21:42] Building go targets for linux/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
+++ [1108 04:21:50] Building go targets for linux/amd64:
    ./vendor/github.com/go-bindata/go-bindata/go-bindata
+++ [1108 04:21:51] Building go targets for linux/amd64:
    cmd/kube-proxy
    cmd/kube-apiserver
    cmd/kube-controller-manager
    cmd/kubelet
    cmd/kubeadm
    cmd/kube-scheduler
    vendor/k8s.io/kube-aggregator
    vendor/k8s.io/apiextensions-apiserver
    cluster/gce/gci/mounter
stat /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/bazel-bin: no such file or directory
+++ [1108 04:24:03] Building go targets for linux/amd64:
    cmd/kube-proxy
    cmd/kubeadm
    cmd/kubelet
stat /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/bazel-bin: no such file or directory
+++ [1108 04:24:26] Building go targets for linux/amd64:
    cmd/kubectl
stat /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/bazel-bin: no such file or directory
+++ [1108 04:24:36] Building go targets for linux/amd64:
    cmd/gendocs
    cmd/genkubedocs
    cmd/genman
    cmd/genyaml
    cmd/genswaggertypedocs
    cmd/linkcheck
    vendor/github.com/onsi/ginkgo/ginkgo
    test/e2e/e2e.test
    cluster/images/conformance/go-runner
stat /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/bazel-bin: no such file or directory
+++ [1108 04:26:19] Building go targets for linux/amd64:
    cmd/kubemark
    vendor/github.com/onsi/ginkgo/ginkgo
    test/e2e_node/e2e_node.test
+++ [1108 04:27:06] Syncing out of container
+++ [1108 04:27:26] Starting tarball: client linux-amd64
+++ [1108 04:27:26] Building tarball: manifests
+++ [1108 04:27:26] Building tarball: src
+++ [1108 04:27:26] Waiting on tarballs
tar: Removing leading `/' from member names
tar: Removing leading `/' from member names
tar: Removing leading `/' from member names
tar: Removing leading `/' from member names
+++ [1108 04:27:35] Building tarball: node linux-amd64
+++ [1108 04:27:35] Building images: linux-amd64
+++ [1108 04:27:36] Starting docker build for image: kube-apiserver-amd64
+++ [1108 04:27:36] Starting docker build for image: kube-controller-manager-amd64
+++ [1108 04:27:36] Starting docker build for image: kube-scheduler-amd64
+++ [1108 04:27:36] Starting docker build for image: kube-proxy-amd64
+++ [1108 04:27:36] Building conformance image for arch: amd64
WARNING: Could not store access token in cache: database is locked
WARNING: Could not store access token in cache: database is locked
WARNING: Could not store access token in cache: database is locked
+++ [1108 04:28:05] Deleting docker image k8s.gcr.io/kube-scheduler-amd64:v1.20.0-beta.1.257_2570a9d94ced19
+++ [1108 04:28:06] Deleting docker image k8s.gcr.io/kube-controller-manager-amd64:v1.20.0-beta.1.257_2570a9d94ced19
+++ [1108 04:28:06] Deleting docker image k8s.gcr.io/kube-apiserver-amd64:v1.20.0-beta.1.257_2570a9d94ced19
+++ [1108 04:28:10] Deleting docker image k8s.gcr.io/kube-proxy-amd64:v1.20.0-beta.1.257_2570a9d94ced19
+++ [1108 04:28:15] Deleting conformance image k8s.gcr.io/conformance-amd64:v1.20.0-beta.1.257_2570a9d94ced19
+++ [1108 04:28:15] Docker builds done
+++ [1108 04:28:15] Building tarball: server linux-amd64
+++ [1108 04:29:14] Building tarball: final
+++ [1108 04:29:15] Starting tarball: test linux-amd64
+++ [1108 04:29:15] Waiting on test tarballs
+++ [1108 04:29:56] Building tarball: test portable
INFO Using build version: v1.20.0-beta.1.257+2570a9d94ced19 
INFO Latest version is v1.20.0-beta.1.257+2570a9d94ced19 
INFO Checking bucket k8s-release-kubernetes for write permissions 
FATA Failed to run:: check release bucket access: fetching gcloud credentials, try running "gcloud auth application-default login": dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information. 

real	9m33.184s
user	2m24.579s
sys	0m14.016s
```

**Comment by justaugustus** [comment]
There's still work to do here, but at this point it's starting to touch some of the areas @saschagrunert is working on in https://github.com/kubernetes/release/pull/1692, so I'd like to merge and iterate.

/assign @saschagrunert @hasheddan @xmudrii @cpanato 
cc: @kubernetes/release-engineering

File: pkg/build/ci.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+			existErrors = append(existErrors, existErr)
+		}
+
+		if i == len(gcsBuildPaths)-1 && len(existErrors) == 0 {
+			logrus.Infof("Build already exists. Exiting...")
+			return true, nil
+		}
+	}
+
+	logrus.Infof("The following error(s) occurred while looking for a build: %v", existErrors)
```
Comment (saschagrunert) [inline pkg/build/ci.go:174]: What is the reason that we do not return the errors (maybe stacked/concatenated) `if len(existErrors) > 0`?

**Review by saschagrunert — APPROVED** [review]
/lgtm

Just one question, but I guess this can be done/resolved as follow-up.

**Pull request merged**